File: method_dispatch_bench.js
#!/usr/bin/env node
// Micro-benchmark: generic function dispatch (JavaScript version)
// Basic dispatch performance test using a dispatch table

// Dispatch table
const dispatchTable = {
  number: (x) => ["number", x],
  string: (x) => ["string", x.length],
  object: (x) => ["object", Object.keys(x).length],
  boolean: (x) => ["boolean", x],
  other: (x) => ["other", x],
};

function getTypeKey(x) {
  if (typeof x === "number") {
    return "number";
  } else if (typeof x === "string") {
    return "string";
  } else if (typeof x === "object" && x !== null) {
    return "object";
  } else if (typeof x === "boolean") {
    return "boolean";
  } else {
    return "other";
  }
}

function dispatchGeneric(x) {
  const key = getTypeKey(x);
  return dispatchTable[key](x);
}

// Test data: 2,000,000 calls with 5-type repeating cycle (10× iterations)
const testData = [];
const cycle = [1, "test string", { a: 1, b: 2, c: 3, d: 4, e: 5 }, true, "symbol"];
for (let repeat = 0; repeat < 10; repeat++) {
  for (let i = 0; i < 200000; i++) {
    testData.push(cycle[i % 5]);
  }
}

let globalResult = 0;

function benchmarkDispatch(iterations) {
  const times = [];
  for (let run = 1; run <= iterations; run++) {
    const startTime = performance.now();
    for (let i = 0; i < testData.length; i++) {
      const r = dispatchGeneric(testData[i]);
      if (r) {
        globalResult = (globalResult + 1) % 1000000; // Force side effect
      }
    }
    const elapsed = performance.now() - startTime;
    console.log(`  Run ${run}: ${(elapsed / 1000).toFixed(1)} seconds`);
    times.push(elapsed);
  }
  return times;
}

function runAllBenchmarks() {
  console.log("\n================================");
  console.log("JavaScript Generic Function Dispatch Micro-Benchmark");
  console.log("================================");
  console.log(`Implementation: Node.js ${process.version}`);
  console.log("Test data: 2,000,000 calls over repeating 5-type cycle (10× iterations)");
  console.log("  Type cycle: number -> string -> object -> boolean -> other\n");

  console.log("Warming up interpreter (10,000 calls)...");
  for (let i = 0; i < 10000; i++) {
    dispatchGeneric(testData[(i % 200000)]);
  }
  console.log("Warmup complete.\n");

  console.log("Running dispatch benchmark (10 iterations):");
  benchmarkDispatch(10);

  console.log(`\nInternal state (preventing JIT optimization): ${globalResult}`);
  console.log("\n================================");
  console.log("Benchmark Complete");
  console.log("================================");
}

runAllBenchmarks();

File: MethodDispatchBench.java
import java.util.*;
import java.util.function.Function;

public class MethodDispatchBench {
    // Dispatch table using a map
    static Map<String, Function<Object, Object>> dispatchTable = new HashMap<>();

    static {
        dispatchTable.put("number", x -> new Object[]{"number", x});
        dispatchTable.put("string", x -> new Object[]{"string", ((String) x).length()});
        dispatchTable.put("object", x -> new Object[]{"object", ((HashMap) x).size()});
        dispatchTable.put("boolean", x -> new Object[]{"boolean", x});
        dispatchTable.put("other", x -> new Object[]{"other", x});
    }

    static String getTypeKey(Object x) {
        if (x instanceof Integer || x instanceof Long || x instanceof Double) {
            return "number";
        } else if (x instanceof String) {
            return "string";
        } else if (x instanceof HashMap) {
            return "object";
        } else if (x instanceof Boolean) {
            return "boolean";
        } else {
            return "other";
        }
    }

    static Object dispatchGeneric(Object x) {
        String key = getTypeKey(x);
        Function<Object, Object> fn = dispatchTable.get(key);
        return fn.apply(x);
    }

    static void benchmarkDispatch(int iterations, List<Object> testData) {
        for (int run = 1; run <= iterations; run++) {
            long startTime = System.nanoTime();
            int result = 0;
            for (Object item : testData) {
                Object r = dispatchGeneric(item);
                if (r != null) result++;
            }
            long elapsedNs = System.nanoTime() - startTime;
            double elapsedMs = elapsedNs / 1_000_000.0;
            double elapsedSec = elapsedMs / 1000.0;
            System.out.printf("  Run %d: %.1f seconds%n", run, elapsedSec);
        }
    }

    public static void main(String[] args) {
        System.out.println("\n================================");
        System.out.println("Java Generic Function Dispatch Micro-Benchmark");
        System.out.println("================================");
        System.out.println("Implementation: " + System.getProperty("java.version"));
        System.out.println("Test data: 200,000 calls over repeating 5-type cycle");
        System.out.println("  Type cycle: number -> string -> object -> boolean -> other\n");

        // Create test data
        List<Object> testData = new ArrayList<>();
        Object[] cycle = new Object[]{
            1,
            "test string",
            new HashMap<String, Integer>() {{
                put("a", 1); put("b", 2); put("c", 3); put("d", 4); put("e", 5);
            }},
            true,
            "symbol"
        };
        for (int i = 0; i < 200000; i++) {
            testData.add(cycle[i % 5]);
        }

        System.out.println("Warming up JIT compiler (20,000 calls)...");
        for (int i = 0; i < 20000; i++) {
            dispatchGeneric(testData.get(i % testData.size()));
        }
        System.out.println("Warmup complete.\n");

        System.out.println("Running dispatch benchmark (3 iterations):");
        benchmarkDispatch(3, testData);

        System.out.println("\n================================");
        System.out.println("Benchmark Complete");
        System.out.println("================================");
    }
}
